File: cpp/47_distinct_prime_factors.cpp
#include <iostream>
#include <vector>
#include <set>

bool is_prime(int n){
    if (n == 2) return true;
    for (int i = 2; i * i <= n; i++){
        if (n % i == 0) return false;
    }
    return true;
}

int count_distinct_prime_factors(int n, std::vector<int> primes){
    std::set<int> factors;
    // iterate over ordered primes
    for (int prime : primes){
        // add prime factors to set
        while (n % prime == 0){
            factors.insert(prime);
            n /= prime;
        }
        // all prime factors found
        if (n == 1) break;
    }
    return factors.size();
}

int distinct_primes_factors(){
    // ordered list of primes for checking prime factors
    std::vector<int> primes;
    int n = 2, streak = 0;
    while (streak < 4){
        if (!is_prime(n)){
            if (count_distinct_prime_factors(n, primes) == 4){
                streak++;
            }
            else streak = 0;
        }
        else{
            primes.push_back(n);
            streak = 0;
        }
        n++;
    }
    return n - streak;
}

int main(){
    std::cout << distinct_primes_factors() << "\n";
}

File: java/_47_DistinctPrimeFactors.java
import java.util.*;

public class _47_DistinctPrimeFactors{
    
    static HashSet<Integer> primes = new HashSet<>();
    
    public static boolean isPrime(int n){
        for (int i = 2; i*i < n; i++){
            if (n % i == 0)
                return false;
        }
        return true;
    }
    
    public static HashSet<Integer> getDistinctPrimeFactors(int n){
        HashSet<Integer> factors = new HashSet<>();
        Iterator<Integer> iter = primes.iterator();
        while (n > 1 && iter.hasNext()){
            int p = iter.next();
            while (n % p == 0){
                factors.add(p);
                n /= p;
            }
        }
        return factors;
    }
    
    public static void main(String[] args){
        
        int n = 2;
        int lastConsecutiveInt = 0;
        int streak = 0;
        
        // The performance of this approach is impacted by the initial primality check for each n.
        // To remedy this, a sieve could be initialized, but there is no immediately apparent upper bound,
        // meaning that the sieve may need to be recalculated with a higher bound if no results are found.
        
        while (lastConsecutiveInt == 0){
            if (!isPrime(n)){
                HashSet<Integer> factors = getDistinctPrimeFactors(n);
                
                if (factors.size() == 4){
                    // List<Integer> factorsList = new ArrayList<>(factors);
                    // Collections.sort(factorsList);
                    // System.out.println(n + ": " + factors.size() + " " + factorsList);
                    streak++;
                }
                else{
                    streak = 0;
                }
            }
            else{ // cache primes for faster lookup
                primes.add(n);
                streak = 0;
            }
            
            if (streak == 4){
                lastConsecutiveInt = n;
            }
            n++;
        }
        
        System.out.println("First of four consecutive integers with four distinct prime factors: " + (lastConsecutiveInt - 3)); // 134043
        // 134043: 3, 7, 13, 491
        // 134044: 2, 23, 31, 47
        // 134045: 5, 17, 19, 83
        // 134046: 2, 3, 11, 677
    }
}